### PR TITLE
TLS server support SNI based certificate selection

### PR DESCRIPTION
**Pull request by cspiel1**
tls: support multiple TLS server certificates
    
    Multiple certificates combined with private key and host name for the host name
    verification can be added with a new function `tls_add_certf()`.
    
    For incoming SIP requests a certificate is selected by SNI in the TLS client
    hello before `SSL_accept()` is called.

e.g. use case: incoming OPTIONS from SIP servers for client is alive polling

TODO: maybe in follow up PRs in order to keep this PR small.
- cleanup/combine with
```
commit 21f36e31522b8d2106e6c1daf64ea0eba626f79c
Author: Christoph Huber <c.huber@commend.com>
Date:   Mon Apr 12 09:45:50 2021 +0200

    transp: add and use different client certificates
```
- The names `tls_set_verify_server()` and  `tls_set_verify_client()` are not very proper. Maybe we could add a deprecated logging/doxygen and add new functions:
  - `tls_conn_enable_verify_peer(struct tls_conn *tc, const char *host)` with optional `host` parameter and
  - `tls_trust_all(struct tls *tls)`

### CODE REVIEW

File: src/tls/openssl/tls.c [hunk 1]
```diff
@@ -1790,3 +1810,377 @@ SSL_CTX *tls_ssl_ctx(const struct tls *tls)
...
+	int err;
+
+	if (!bio || !sni)
+		return EINVAL;
+
+	pl.l = BIO_get_mem_data(bio, (char **)&pl.p);
+	if (!pl_isset(&pl))
+	    return EINVAL;
+
+	/* skip record header, handshake header, client version and random */
```
[the remark this answers is not in the document]
Comment (cspiel1) [reply src/tls/openssl/tls.c:1893]: > is it possible to do the TLS message parsing in OpenSSL instead ?
> does it work with TLS 1.2 and TLS 1.3 ?

I searched in the openssl code for a while and couldn't find an adequate function. Today, we found the API function:
`SSL_CTX_set_client_hello_cb()`

I try to learn from the apache source code how they do cert selection via SNI. I set this PR to draft so far.

**Comment by cspiel1** [comment]
Apache copied code from `openssl/test/helpers/handshake.c` for parsing the SNI from the client hello.

The last commit in this PR also uses `SSL_CTX_set_client_hello_cb()` and `SSL_client_hello_get0_ext()` to do most of the parsing. This replaces most of the parsing code.

I guess it works only for TLS 1.3 because in man page of `SSL_client_hello_isv2` there is:
```
The SSLv2 format has substantial differences from the normal SSLv3 format, including using three bytes per cipher suite, and not allowing extensions.
```

Note: I have to test this before we can proceed.

**Comment by alfredh** [comment]
https://stackoverflow.com/questions/5113333/how-to-implement-server-name-indication-sni


it would be nice to implement this without any manual parsing of packets.

we support TLS v1.2 and v1.3
SSLv2/3 is deprecated.

**Comment by cspiel1** [comment]
Thanks for this hint Alfred!
Currently we have some release pressure. I'll try to rewrite this PR as soon as possible.

File: src/tls/openssl/sni.c [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+#include <re_list.h>
+#include <re_hash.h>
+
+
+enum {
+	HANDSH_PROT_SKIP    = 5,
+	CONTENT_TYPE_HANDSH = 22,
+	HANDSH_CLIENT_HELLO = 1,
+	RECORD_LAYER_SKIP   = 4,
+};
```
Comment (alfredh) [inline src/tls/openssl/sni.c:33]: are these used ?

File: src/tls/openssl/sni.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+static int pl_bin_u8(struct pl *pl, uint8_t *len)
+{
+	if (!pl_isset(pl))
+		return EINVAL;
+
+	*len = (uint8_t) (*pl->p);
+	pl_advance(pl, 1);
+	return 0;
+}
```
Comment (alfredh) [inline src/tls/openssl/sni.c:45]: struct pl is designed for operating on strings.

for this it is probably better to use mbuf

File: src/tls/openssl/tls.c [hunk 4]
```diff
@@ -143,7 +154,7 @@ static int keytype2int(enum tls_keytype type)
 
 #if OPENSSL_VERSION_NUMBER >= 0x10100000L && \
 	!defined(LIBRESSL_VERSION_NUMBER)
-static int verify_handler(int ok, X509_STORE_CTX *ctx)
+int tls_verify_handler(int ok, X509_STORE_CTX *ctx)
```
Comment (alfredh) [inline src/tls/openssl/tls.c:157]: I would prefer handlers to be static.

File: src/tls/openssl/tls_tcp.c [hunk 5]
```diff
@@ -241,7 +241,7 @@ static bool recv_handler(int *err, struct mbuf *mb, bool *estab, void *arg)
 			*err = tls_accept(tc);
 		}
 
-		DEBUG_INFO("state=0x%04x\n", SSL_state(tc->ssl));
+		DEBUG_INFO("state: %s\n", SSL_state_string_long(tc->ssl));
```
Comment (alfredh) [inline src/tls/openssl/tls_tcp.c:244]: perhaps move to separate PR ?

**Comment by cspiel1** [comment]
Thanks! First I try to rewrite this like you already mentioned: https://github.com/baresip/re/pull/596#issuecomment-1330240156

File: src/tls/openssl/tls.c (hunk 4 above)
Comment (cspiel1) [reply src/tls/openssl/tls.c:157]: Okay.

The perfect name for a function in `tls.c` that turns on SSL_VERIFY_PEER for a TLS client and sets a verify handler would be: `void tls_set_verify_client(tls)`

But there is already a function with this name that does the opposite. It's sets the handler `verify_trust_all`. :-(

**Comment by cspiel1** [comment]
Separated: https://github.com/baresip/re/pull/613

**Comment by cspiel1** [comment]
TODO: Test the change to the SNI handler.

**Comment by alfredh** [comment]
I would like to suggest the following:

1. Create a new patch this in re
2. Create a matching test in retest


The test can be a back-to-back test with a TLS connection
from Client to Server, that is using SNI.

If you want you can close this PR and make a new one.

**Comment by cspiel1** [comment]
New PR is not necessary. I re-based and squashed the commits.

Now I'll start to work on the retest.

**Comment by alfredh** [comment]
some general comments:

- use `const char *` instead of `struct pl` in the public API
- in for loops, move declaration of `i` into for loop
- in all openssl Free functions, check if they can handle a NULL pointer
- in all places where openssl return error, call `ERR_clear_error()`
- global symbols in the tls module should have prefix: `tls_`
- move sni.h into tls.h
- add doxygen comment to public functions
-

File: src/tls/openssl/tls.c [hunk 6]
```diff
@@ -1779,3 +1825,131 @@ SSL_CTX *tls_ssl_ctx(const struct tls *tls)
...
+		}
+
+		if (!sk_X509_push(uc->chain, ca)) {
+			err = ENOMEM;
+			goto out;
+		}
+	}
+
+	BIO_free(bio);
+	bio = BIO_new_file(certf, "r");
```
Comment (alfredh) [inline src/tls/openssl/tls.c:1891]: BIO_new_file is called twice, is it needed ?

Can you use the same BIO for cert and private key ?

**Comment by cspiel1** [comment]
> * in all openssl Free functions, check if they can handle a NULL pointer

Yes they have.
- ASN1 free functions --> checked in the source code
- `X509_free` and `EVP_PKEY_free` --> checked in man page
- `sk_X509_pop_free` --> checked in the code

> * in all places where openssl return error, call `ERR_clear_error()`
Okay, I'll put it in the "out:" section like it is done in other function in the tls module. It doesn't hurt if it called too often.

File: src/tls/openssl/tls.c (hunk 6 above)
Comment (cspiel1) [reply src/tls/openssl/tls.c:1891]: I already read about `BIO_reset()`. But can't remember the result. I will test it when I have the retest.

File: src/tls/openssl/tls.c [hunk 7]
```diff
@@ -266,6 +290,7 @@ int tls_alloc(struct tls **tlsp, enum tls_method method, const char *keyfile,
 		}
 	}
 
+	tls_enable_sni(tls);
```
Comment (alfredh) [inline src/tls/openssl/tls.c:293]: should SNI be enabled by default, also for clients ?

File: src/tls/openssl/tls.c [hunk 8]
```diff
@@ -1779,3 +1804,168 @@ SSL_CTX *tls_ssl_ctx(const struct tls *tls)
...
+			goto out;
+		}
+
+		if (!sk_X509_push(uc->chain, ca)) {
+			err = ENOMEM;
+			goto out;
+		}
+	}
+
+	BIO_reset(bio);
```
Comment (alfredh) [inline src/tls/openssl/tls.c:1882]: is this line needed ?

File: src/tls/openssl/tls.c (hunk 7 above)
Comment (cspiel1) [reply src/tls/openssl/tls.c:293]: We can enable it in first call of `tls_add_certf()`.

File: src/tls/openssl/tls.c (hunk 8 above)
Comment (cspiel1) [reply src/tls/openssl/tls.c:1882]: Yes. See https://github.com/baresip/re/pull/596#discussion_r1046782443!

These steps are done in `tls_add_certf()`:
- first read certificate
- read potentially following CAs
- reset the BIO and read the key anywhere in the file
